### ISSUE TITLE
Add support for structured logging for diagnostics

### ISSUE THREAD

**Issue by caioproiete** [Excel-DNA/ExcelDna#277]
Excel-DNA's [Diagnostic Logging](https://github.com/Excel-DNA/ExcelDna/wiki/Diagnostic-Logging) feature allows the developer to configure a `TraceListener` that will be used by Excel-DNA to write diagnostic log messages, which can then be forwarded to any logging pipeline such as [Serilog](https://serilog.net) or [NLog](https://github.com/NLog/NLog), for example, in order to consume / analyse these logs outside of the user's machine.

Unfortunately, the standard `Trace` feature that comes with .NET only does simple text logging and does not provide [structured logging](https://nblumhardt.com/2016/06/structured-logging-concepts-in-net-series-1/) capabilities, and important information is lost by the time messages reach the logging pipeline.

![image](https://user-images.githubusercontent.com/177608/68549501-fc581a00-03ce-11ea-82bd-6ba998a1c210.png)

It would be great if Excel-DNA moved away from `Trace` and added support for structured logging within its diagnostic logging, and provided an extension point for the developer to attach structured logging providers at run-time.

Ideally it should use [Microsoft.Extensions.Logging](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging?view=dotnet-plat-ext-3.0) which is the current standard for modern .NET apps.

**Comment by n1l** [comment]
What do you think about ETW?

**Comment by caioproiete** [comment]
@n1l Are you suggesting we write logs directly to ETW from within Excel-DNA?

It's definitely possible, but I see ETW more as a possible "_destination_" for the logs, instead of a "_source_" for logs.

I think Excel-DNA should _not_ decide where to write logs, and instead provide a mechanism that allows the user to plug-in any .NET logging library that he/she wants, such as [Serilog](https://serilog.net) or [NLog](https://github.com/NLog/NLog), for example, and the logging library then decides where the logs are written based on configuration provided by the Excel Add-In.

That way, if the user wants logs to end up in ETW, they can do so via the integration provided by the logging library they are using.

e.g.
![image](https://user-images.githubusercontent.com/177608/72359137-21655380-36c4-11ea-8cda-018da696b113.png)

Does that make sense?

The `Trace` mechanism used in Excel-DNA today already provides a way to get the pipeline above by implementing a custom `TraceListener` (see [`ExcelDna.Diagnostics.Serilog`](https://github.com/caioproiete/exceldna-diagnostics-serilog) for example) that forwards logs to a logging library, but everything is written in plain text without preserving field names and values.

I think [Microsoft.Extensions.Logging](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging?view=dotnet-plat-ext-3.0) _at the moment_ would be the best candidate for the logging API _glue_, replacing the `Trace`.

**Comment by n1l** [comment]
Yes It's absolutely does, and I definitely need to try it out, thx!
